module-8 deliverables/advancedNFT/contracts/AdvancedNFT.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Multicall.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+contract AdvancedNFT is ERC721, Multicall, Ownable {
+
+    // State machine for minting phases
+    enum SaleState { INACTIVE, PRESALE, PUBLIC, SOLD_OUT }
+    SaleState public saleState;
+    bytes32 public merkleRoot;
+    // Mapping for commit-reveal
+    mapping(address => bytes32) public commitments;
+    mapping(address => uint256) public commitBlock;
+    mapping(address => bool) public hasMintedMapping; 
+
+    // NFT supply and pricing, and address for withdraw
+    uint256 public constant MAX_SUPPLY = 100; // Changed from 10000 to 100
+    uint256 public constant PRESALE_LIMIT = 10; // New constant for presale limit
+    uint256 public presaleMinted; // Track presale mints separately
+    uint256 public totalSupply;
```

Comment:
Just be careful adding state variables like this. They need to be modified by any function that will increase or decrease the total supply. As it stands this contract does not decrease the total supply when tokens are burned. In this contract you could override the _update function or provide your own version of the mint and burn functions that will update the totalSupply.